core/include/detray/navigation/direct_navigator.hpp

```diff
@@ -0,0 +1,366 @@
+                                        const context_type &ctx) const {
+        if (navigation.is_complete()) {
+            navigation.m_heartbeat = false;
+            return;
+        }
+
+        navigation.m_heartbeat = true;
+        navigation.update_candidate(!navigation.is_init());
+        update_intersection(track, navigation, cfg, ctx);
+
+        return;
+    }
+
+    template <typename track_t>
+    DETRAY_HOST_DEVICE inline bool update(const track_t &track,
+                                          state &navigation,
+                                          const navigation::config &cfg,
+                                          const context_type &ctx = {}) const {
+
+        if (navigation.is_complete()) {
+            navigation.m_heartbeat = false;
+            return true;
+        }
+
+        assert(!navigation.get_target_barcode().is_invalid());
+        update_intersection(track, navigation, cfg, ctx);
+
+        if (navigation.is_on_surface(navigation.target(), cfg)) {
+            navigation.m_status = (navigation.target().sf_desc.is_portal())
+                                      ? navigation::status::e_on_portal
+                                      : navigation::status::e_on_module;
+            assert(navigation.is_on_surface(navigation.target(), cfg));
+            navigation.next();
+            navigation.update_candidate(true);
+            assert(navigation.is_on_surface(navigation.current(), cfg));
+
+            if (!navigation.is_complete()) {
+                update_intersection(track, navigation, cfg, ctx);
+            }
+
+            return true;
+        } else {
+            // Otherwise the track is moving towards a surface
+            navigation.m_status = navigation::status::e_towards_object;
+
+            return true;
+        }
+    }
+
+    template <typename track_t>
+    DETRAY_HOST_DEVICE inline void update_intersection(
+        const track_t &track, state &navigation, const navigation::config &cfg,
+        const context_type &ctx = {}) const {
+
+        const auto &det = navigation.detector();
+        const auto sf = tracking_surface{det, navigation.target().sf_desc};
+
+        const bool res =
+            sf.template visit_mask<intersection_update<ray_intersector>>(
+                detail::ray<algebra_type>(
+                    track.pos(),
+                    static_cast<scalar_type>(navigation.direction()) *
+                        track.dir()),
+                navigation.target(), det.transform_store(), ctx,
+                sf.is_portal() ? darray<scalar_type, 2>{cfg.min_mask_tolerance,
+                                                        cfg.max_mask_tolerance}
+                               : darray<scalar_type, 2>{cfg.min_mask_tolerance,
+                                                        cfg.max_mask_tolerance},
+                static_cast<scalar_type>(cfg.mask_tolerance_scalor),
+                static_cast<scalar_type>(cfg.overstep_tolerance));
```

Comment:
Tha mask tolerance and overstepping tolerance are not needed for direct navigation, I believe
```suggestion
                darray<scalar_type, 2>{cfg.min_mask_tolerance,
                                                         cfg.min_mask_tolerance},
               0.f,
               -detail::invalid_value<scalar_type>());
```

Reply:
Without mask tolerance, the surfaces can get missed during the backward propagation.

The overstep tolerance could be any values technically but I worry the situation where the navigation is trapped in an infinite loop.

Reply:
The direct navigator should not miss a surface when the intersection is outside or it would run precisely into the same problems the current navigator does. Or am I missing something?

> The overstep tolerance could be any values technically but I worry the situation where the navigation is trapped in an infinite loop.

I am sorry, I cannot follow on how it could get stuck this way. Can you explain what you mean, please?

Reply:
Ah, OK, that is interesting! Can you show me how to reproduce this?

Reply:
And in our geometries during tracking, is there a configuration you found that I can try?

Reply:
No It does not have in our particular case. 

As I said, I am trying to make it work on arbitrary cases like Geant4 navgiation does

Reply:
> And in our geometries during tracking, is there a configuration you found that I can try?

Just make a telescope geometry if you want to try. Do you need a exact number and codes? No way that I am going to write it for you. You just don't like this or what??

Reply:
I was asking, because I would like to debug this case and was wondering if you had already encountered this and could give me a starting point... that is all, I am sorry

Reply:
I don't think you would ever try it on your own.  you just want to prove that I have not encountered this actually so shut down my argument, don't you?

As I said, set it to -MAX with the configuration object if you want.

Reply:
No, this problem makes sense to me and I had not thought of it before. I am genuinely sorry if you feel that way

Reply:
:clown_face: